File: Common/Actor/src/common/actor/Collectable.java
/**
 * @author Julian Sender
 */
package common.actor;

import java.util.List;

/**
 * Abstract class as blueprint for all collectables
 */
abstract public class Collectable extends Actor {

    public static final String collected = "I am collected, remove me :)";
    private             Actor  collector = null;

    /**
     * Extending Actor and needing to give Actor the picture
     * Then following many overloadings of constructor
     * @param pictureFileName path of desired picture
     */
    protected Collectable( String pictureFileName ) {
        super( pictureFileName );
    }

    protected Collectable( double x, double y, int delay, double scale, String... pictureFilePaths ) {
        super( x, y, delay, pictureFilePaths );
        this.scaleImage( scale );
    }

    public Collectable( double x, double y, String invisiblePictures ) {super( invisiblePictures, x, y );}

    /**
     * Notifys ovservers if a collectable was collected
     * @param collector param for new collected to notify observers
     */
    public void wasCollected( Actor collector ) {
        this.collector = collector;
        this.setChanged();
        this.notifyObservers( collected );
    }

    /**
     * GETTER
     * @return Returns a collector-instance
     */
    public Actor getCollector() {
        return collector;
    }
}

File: Common/Actor/src/common/actor/Collectable.kt
package common.actor

/**
 * @author Julian Sender
 */
/**
 * Abstract class as blueprint for all collectables
 */
abstract class Collectable : Actor {
    /**
     * GETTER
     * @return Returns a collector-instance
     */
    var collector: Actor? = null
        private set

    /**
     * Extending Actor and needing to give Actor the picture
     * Then following many overloadings of constructor
     * @param pictureFileName path of desired picture
     */
    protected constructor(pictureFileName: String) : super(pictureFileName) {}
    protected constructor(x: Double,
                          y: Double,
                          delay: Int,
                          scale: Double,
                          vararg pictureFilePaths: String) : super(pictureFilePaths.toList(), x, y, delay.toDouble()) {
        scaleImage(scale)
    }

    constructor(x: Double, y: Double, invisiblePictures: String?) : super(invisiblePictures!!,
                                                                          x,
                                                                          y) {
    }

    /**
     * Notifys ovservers if a collectable was collected
     * @param collector param for new collected to notify observers
     */
    open fun wasCollected(collector: Actor) {
        this.collector = collector
        setChanged()
        this.notifyObservers(collected)
    }

    companion object {
        const val collected = "I am collected, remove me :)"
    }
}
